cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,328 @@
+    kept_cycles = []
+
+    for ind, vec in enumerate(kak_vectors):
+        # dists = KAK_vector_infidelity(vec, KAK_mesh)
+        # The L2 distance is an upper bound to the locally invariant distance,
+        # but it's much faster to compute.
+        dists = np.sqrt(np.sum((KAK_mesh - vec)**2, axis=-1))
+        close = (dists < max_dist).nonzero()[0]
+        assert close.shape[0] in (0, 1), f'shape: {close.shape}'
+        cycles_for_gate = tuple(
+            (k_0[ind], k_1[ind]) for k_0, k_1 in local_unitary_pairs)
+
+        kept = False
+        for mesh_ind in close:
+            if mesh_ind not in tabulation:
+                tabulation[mesh_ind] = cycles_for_gate
+                kept = True
+
+        if kept:
+            kept_kaks.append(vec)
+            kept_cycles.append(cycles_for_gate)
+
+    return kept_kaks, kept_cycles
+
+
+def gate_product_tabulation(base_gate: np.ndarray,
+                            max_infidelity: float,
+                            verbose: bool = False,
+                            include_warnings: bool = True) -> GateTabulation:
+    r"""Generate a GateTabulation for a base two qubit unitary.
+
+    Args:
+        base_gate: The base gate of the tabulation.
+        max_infidelity: Sets the desired density of tabulated product unitaries.
+            The typical nearest neighbor Euclidean spacing (of the KAK vectors)
+            will be on the order of \sqrt(max_infidelity). Thus the number of
+            tabulated points will scale as max_infidelity^{-3/2}.
+        verbose: Whether to print gate tabulation statistics.
+        include_warnings: If True, warn the user if a point in the Weyl
+            chamber has no tabulated points within the desired distance.
+
+    Returns:
+        A GateTabulation object used to compile new two-qubit gates from
+        products of the base gate with 1-local unitaries.
+    """
+    assert 1 / 2 > max_infidelity > 0
+    spacing = np.sqrt(max_infidelity / 3)
+    mesh_points = weyl_chamber_mesh(spacing)
+
+    num_samples = mesh_points.shape[0] * 50
+
+    # include the base gate itself
+    kak_vecs = [kak_vector(base_gate, check_preconditions=False)]
+    sq_cycles: List[Tuple[_SingleQubitGatePair, ...]] = [()]
+
+    # Tabulate gates that are close to gates in the mesh
+    u_locals_0 = random_qubit_unitary(num_samples)
+    u_locals_1 = random_qubit_unitary(num_samples)
+
+    u_locals_for_gate: Dict[int, Tuple[_SingleQubitGatePair, ...]] = {}
+    tabulation_cutoff = 0.5 * spacing
+    out = _tabulate_KAK_vectors(u_locals_for_gate, base_gate, tabulation_cutoff,
+                                mesh_points, (u_locals_0, u_locals_1))
+    kak_vecs.extend(out[0])
+    sq_cycles.extend(out[1])
+
+    # Will be used later for getting missing KAK vectors.
+    kak_vecs_single = np.array(kak_vecs[1:])
+    sq_cycles_single = list(sq_cycles[1:])
+
+    if verbose:
+        # coverage: ignore
+        print(f'fraction satisfied with 2 gates'
```

Comment:
This kind of output is not really appropriate within a method inside of a large framework. E.g. if you called the quantum service and saw output about 'fraction satisfied with 2 gates' you'd be pretty confused by that. This information should be e.g. a method on the type returned by the method. The fact that you had to ignore it for coverage purposes is indicative of this.

Reply:
Done.